src/rust/engine/src/nodes.rs

```diff
@@ -902,34 +902,49 @@ impl Task {
           .and_then(|edges| {
-            edges
-              .entry_for(&dependency_key)
-              .cloned()
-              .ok_or_else(|| match get.input_type {
-                Some(ty) if externs::is_union(ty) => {
-                  let value = externs::type_for_type_id(ty).into_object();
-                  match externs::call_method(
-                    &value,
-                    "non_member_error_message",
-                    &[externs::val_for(&get.input)],
-                  ) {
-                    Ok(err_msg) => throw(&externs::val_to_str(&err_msg)),
-                    // If the non_member_error_message() call failed for any reason,
-                    // fall back to a generic message.
-                    Err(_e) => throw(&format!(
-                      "Type {} is not a member of the {} @union",
-                      get.input.type_id(),
-                      ty
-                    )),
-                  }
+            edges.entry_for(&dependency_key).cloned().ok_or_else(|| {
+              if externs::is_union(get.input_type) {
+                let value = externs::type_for_type_id(get.input_type).into_object();
+                match externs::call_method(
+                  &value,
+                  "non_member_error_message",
+                  &[externs::val_for(&get.input)],
+                ) {
+                  Ok(err_msg) => throw(&externs::val_to_str(&err_msg)),
+                  // If the non_member_error_message() call failed for any reason,
+                  // fall back to a generic message.
+                  Err(_e) => throw(&format!(
+                    "Type {} is not a member of the {} @union",
+                    get.input.type_id(),
+                    get.input_type
+                  )),
                 }
-                _ => throw(&format!(
-                  "{:?} did not declare a dependency on {:?}",
-                  entry, dependency_key
-                )),
-              })
+              } else {
+                let parent_rule_func = match &*entry {
+                  rule_graph::Entry::WithDeps(entry_with_deps) => match entry_with_deps.rule() {
+                    Some(ref rule) => match rule {
+                      tasks::Rule::Task(ref task) => Some(task.func),
+                      _ => None,
+                    },
+                    None => None,
+                  },
+                  _ => None
+                };
+                let rule_context = match parent_rule_func {
+                  Some(func) => format!(
+                    " See the rule {}() in {} (line {}).",
+                    func.name(), func.module(), func.line_number()
+                  ),
+                  _ => "".to_string()
+                };
+                throw(&format!(
+                  "Invalid input to `Get({}, {})`. Expected an object with type `{}`, but got `{:?}` with type `{}` instead.{}",
```

Comment:
Mentioned offline, but: it seems like this could fail in the `Get` constructor for non-unions? Perhaps it is clearer to have the error handling centralized though.

Reply:
Agreed. I'm working on this rewrite. The benefit is that the stacktrace will include where the problematic Get is coming from.